Provide non blocking API

rnestler opened the pull request.

rnestler committed:
Provide non blocking API for measurements

That way the user can choose to just start a measurement and read the
results later.

rnestler committed:
Pass the delay object only to functions needing it

Most of the time delays need to be used all over in the system, thus we
can't just take ownership of it.

dbrgn commented on src/lib.rs:
```diff
@@ -505,9 +502,13 @@ where
     ///
     /// This is a blocking function call.
-    pub fn measure(&mut self, mode: PowerMode) -> Result<Measurement, Error<E>> {
+    pub fn measure<D: DelayUs<u16>>(
+        &mut self,
+        mode: PowerMode,
+        delay: &mut D,
+    ) -> Result<Measurement, Error<E>> {
```
You could also use the impl trait syntax:

```suggestion
    pub fn measure(
        &mut self,
        mode: PowerMode,
        delay: &mut impl DelayUs<u16>,
    ) -> Result<Measurement, Error<E>> {
```

Might be easier to read, what do you think?

rnestler replied:
Fixed.

dbrgn reviewed (changes requested):
Can you provide multiple impl blocks (common functionality, blocking API, non-blocking API) so we get nicer docs?

rnestler committed:
Split measurement functions into separate impl blocks

That way the documentation is easier to read, since the async and
blocking measurement functions are separate.

Co-authored-by: Danilo Bargen <mail@dbrgn.ch>

dbrgn commented on src/lib.rs:
```diff
@@ -465,6 +465,27 @@ where
+    }
+}
+
+/// Asynchronnous functions for measurement
```
```suggestion
/// Asynchronous functions for starting / reading measurements.
```

dbrgn commented on src/lib.rs:
```diff
@@ -499,6 +515,42 @@ where
+    }
+}
+
+/// Blocking functions for measurement
```
```suggestion
/// Blocking functions for doing measurements.
```

dbrgn reviewed (approved):
Thanks! There are some more things that should be done before the next release:

- Provide an async wakeup function as well
- Add notes about async to README and to crate-level documentation
- Add an example that uses async calls (ideally an embedded one)

However all of them can be added in separate PRs, so let's merge this.

@rnestler can you squash the last two commits before merging?
